### DISCUSSION TITLE
How to use AzureOpenAITextEmbeddingGenerationService.UpsertBatchAsync

### DISCUSSION THREAD

**Discussion by ordinaryAndConfident** [microsoft/semantic-kernel#9846] [Q&A] [answered]
My current confusion is that when I break down a document into many small fragments, using UpsertAsynchronous can lead to excessive database links. How can I solve this problem

**Comment by westey-m** [comment, marked as answer]
@ordinaryAndConfident when considering a chunking strategy it's important to strike a balance on size.  If you create chunks that are too small they may be less likely to be found when doing vector searches.  If the chunks are too big, they can increase token usage when passed to the LLM for context.  The right size will also depend on the type of data that you want to generate embeddings for, so experimenting with different sizes is important.

**Reply by ordinaryAndConfident** [reply]
Now I have encountered a new problem. When the number of documents I need to upload is too large, the efficiency of uploading to CosmosDB Nosql will become particularly low because CosmosDb only supports single-item uploads. Should I change the database?

**Reply by westey-m** [reply]
@ordinaryAndConfident, you could explore uploading documents with multiple threads for CosmosDB to improve efficiency.  This is not supported by the Semantic Kernel implementation, so you would have to develop it, but it could improve speed.

You could certainly consider an alternative database as well, e.g. Azure AI Search supports batch upserts natively.
